js/form_data.ts

```diff
@@ -0,0 +1,173 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import * as domTypes from "./dom_types";
+import { CreateIterableIterator } from "./util";
+import { DenoBlob } from "./blob";
+import { DenoFile } from "./file";
+
+export class FormData implements domTypes.FormData {
+  private data: Array<[string, domTypes.FormDataEntryValue]> = [];
```

Comment:
Also why is this not a `Map`?

Reply:
> My personal opinion is that we should follow the convention of prefacing private members and methods with `_`. In TypeScript, because these are not really private, they impact the type structure. Which means that something else that has a public `data` member is now incompatible structurally, which is undesired. Also, as the TC39 proposal for public and private members becomes a reality, if this is made really private in the future, then there will be a need to refactor and replace the `_` sigil with the private member sigil (`#`) and so having an existing sigil makes that refactoring easier.

If anything we would use a symbol instead: https://github.com/denoland/deno/blob/c4bddc4651f583241193b80634a4f22abb02d582/js/blob.ts#L5-L8

> Also why is this not a `Map`?

Because you can "append" multiple elements under the same name, getAll gets a list of all of them.

Reply:
> My personal opinion is that we should follow the convention of prefacing private members and methods with `_`. In TypeScript, because these are not really private, they impact the type structure. Which means that something else that has a public `data` member is now incompatible structurally, which is undesired. Also, as the TC39 proposal for public and private members becomes a reality, if this is made really private in the future, then there will be a need to refactor and replace the `_` sigil with the private member sigil (`#`) and so having an existing sigil makes that refactoring easier.

If anything we would use a symbol instead: https://github.com/denoland/deno/blob/c4bddc4651f583241193b80634a4f22abb02d582/js/blob.ts#L5-L8

> Also why is this not a `Map`?

Because you can "append" multiple elements under the same name, getAll gets a list of all of them. But yeah, I can change it to use a `Map<string, domTypes.FormDataEntryValue[]>`

Reply:
If we were to go performance-wise, I believe `Array#slice()` (with no arguments) is much faster.

```javascript
getAll(name: string): domTypes.FormDataEntryValue[] {
  const entries = this[dataSymbol].get(name);
  return entries ? entries.slice() : [];
}
```

But then again using the map would be against the implementation of various browsers where FormData#forEach, URLSearchParams#forEach, and others, call the callback strictly by insertion order, which a Map does not allow.